### PR TITLE
Issue 338 sort technician

### PR DESCRIPTION
**Pull request by dtsosie**
https://github.com/MegaMek/mekhq/issues/338

Resolution of above issue. Progress over perfection. This change will notify players if they are about to use a wrong tech, and it will put the right techs per the Tech class (is right (refit target)) at the top.

I think this is best because tbh, I don't assign techs based on their skills. Some repairs take more or less long and I will assign like that. But a new player will assign a tech and only at the end discover they made a mistake. This fixes that issue.

### CODE REVIEW

File: MekHQ/src/mekhq/gui/CampaignGUI.java [hunk 1]
```diff
@@ -1611,7 +1613,11 @@ public void refitUnit(Refit r, boolean selectModelName) {
                         + tech.getMinutesLeft() + "/" + TimePerDay
                         + " minutes";
                 techHash.put(name, tech);
-                techList.add(name);
+                if(tech.isRightTechTypeFor(r)) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/CampaignGUI.java:1616]:
```suggestion
                if (tech.isRightTechTypeFor(r)) {
```

File: MekHQ/src/mekhq/gui/CampaignGUI.java [hunk 2]
```diff
@@ -1622,7 +1628,20 @@ public void refitUnit(Refit r, boolean selectModelName) {
                 return;
             }
 
-            r.setTech(techHash.get(s));
+            Person selectedTech = techHash.get(s);
+
+            if(!selectedTech.isRightTechTypeFor(r)) {
+                if(JOptionPane.NO_OPTION == JOptionPane
+                    .showConfirmDialog(
+                    null,
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/CampaignGUI.java:1636]:
```suggestion
            if (!selectedTech.isRightTechTypeFor(r)) {
                if (JOptionPane.NO_OPTION == JOptionPane.showConfirmDialog(null,
```

File: MekHQ/src/mekhq/gui/CampaignGUI.java [hunk 3]
```diff
@@ -1611,7 +1613,11 @@ public void refitUnit(Refit r, boolean selectModelName) {
                         + tech.getMinutesLeft() + "/" + TimePerDay
                         + " minutes";
                 techHash.put(name, tech);
-                techList.add(name);
+                if(tech.isRightTechTypeFor(r)) {
+                    techList.add(lastRightTech++, name);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/CampaignGUI.java:1617]: If this variable is removed and replaced with a 0, then you get the best techs to refit with first, while any other techs are sorted descending.

Alternatively, you could swap Line 1590 to:
```List<Person> techs = getCampaign().getTechs(false, null, true, true);```
And keep the lastRightTech, which would have the best of the proper type first, then the best of any other type, sorted by their time left.

Both changes allow one to remove the PrimaryRole sort, as why do we need to group the others by role?

**Review by Windchild292 — COMMENTED** [review]
Some style issues with if statements, and a comment on and alternative order for implementation.

File: MekHQ/src/mekhq/gui/CampaignGUI.java (hunk 3 above)
Comment (dtsosie) [reply MekHQ/src/mekhq/gui/CampaignGUI.java:1617]: The idea was minimal change. Inserting in this order maintains the same order the user will see in the rest of the list, just with proper techs promoted to the front. I like the idea of removing PrimaryRole sort though. Back to the salt mines ...

**Review by Windchild292 — COMMENTED** [review]
The remaining time doesn't appear to be sorted by, but otherwise this looks good.

**Comment by dtsosie** [comment]
Well this is decision point time. #338 says they want to redo the whole thing anyway so is this going to be a a step toward that or a quick fix. We're already where it catches the common pain points and warns users if they are about to make a mistake. If we REALLY want to sort by time Person needs to know the skills involved. We will need to remove the primary secondary skill logic from the Tech loop in Campaign.java, incorporate it within Person.java so that the natural sort can know what times to generate. Perhaps overloading with Person.resetTimes(Refit r)? At that point we've signed ourselves up for including refit logic and this new functionality to CampaignTest.java. As this will take me some time to do I'd like feedback since I really don't know what the current course of the platform is and this will require some architecting unlike the quick bug fix I was attempting to do.

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**